### ISSUE TITLE
bqpjson spec - string ids

### ISSUE THREAD

**Issue by ccoffrin** [ccoffrin/bqpjson#22]
Should the current integer ids be replaced with strings?  @spakin

**Comment by ccoffrin** [comment]
The original motivation for using int ids instead of string ids was primarily a question of performance.  Ints are much smaller to store and are fast to hash in hash tables.  This also plays nicely with optimization model standards where the variables are numbered from `0..n-1` or `1..n`.

It is a lucky coincidence that the d-wave qubit names are natively encoded as ints as well.  So the current int-based spec seems to cover all core use cases.  If a new hardware architecture used string-based site ids, that would be a strong motivation for making this change.

**Comment by ccoffrin** [comment]
As a motivational point, thus far I have been thinking of bqpjson as something like a solver independent "bytecode" for solving these kinds of optimization problems.  I have not yet thought about if it could/should be used for encoding a logical versions of a problem.  In the spirit of bytecode it seems to make sense to abstract variable names to integers.

**Comment by spakin** [comment]
> The original motivation for using int ids instead of string ids was primarily a question of performance. Ints are much smaller to store and are fast to hash in hash tables.

Yes, but you can have the solver map strings to/from ids as a pre-/post-processing step to gain the both convenience of strings and the speed of ints.

Consider the following:
```json
{
  "quadratic_terms": [
    {
      "id_tail": 1,
      "id_head": 2,
      "coeff": 0.5
    },
    {
      "id_tail": 1,
      "id_head": 3,
      "coeff": -1.0
    },
    {
      "id_tail": 2,
      "id_head": 3,
      "coeff": -1.0
    }
  ],
  "linear_terms": [
    {
      "id": 1,
      "coeff": 0.5
    },
    {
      "id": 2,
      "coeff": 0.5
    },
    {
      "id": 3,
      "coeff": -1.0
    }
  ]
}
```
That's a Boolean OR.  But which ids correspond to inputs, and which correspond to the output?  Strings to the rescue!
```json
{
  "quadratic_terms": [
    {
      "coeff": 0.5,
      "id_head": "In2",
      "id_tail": "In1"
    },
    {
      "coeff": -1.0,
      "id_head": "Out",
      "id_tail": "In1"
    },
    {
      "coeff": -1.0,
      "id_head": "Out",
      "id_tail": "In2"
    }
  ],
  "linear_terms": [
    {
      "id": "In1",
      "coeff": 0.5
    },
    {
      "id": "In2",
      "coeff": 0.5
    },
    {
      "id": "Out",
      "coeff": -1.0
    }
  ]
}
```

**Comment by ccoffrin** [comment]
This is a good example, but it might be a bit too application specific.  In any case, I can run some tests on the performance impacts of moving from ints to strings for the ids.  If the overhead is negligible in python3, my primary concern will be moot, and I am fine to make this change.
